### ISSUE TITLE
MobileAdvert don't load consecutive rewards video

### ISSUE THREAD

**Issue by rraffo** [Scirra/Construct-bugs#6744]
## Problem description
The first loaded rewarded video is shown correctly. If I load again a rewarded video I'm not able to show it.
I need to load a rewarded video twice to show it.
Mobile APKs are not affected by the problem.

## Attach a .c3p

[admob test.android.debug.apk.zip](https://github.com/Scirra/Construct-bugs/files/10945355/admob.test.android.debug.apk.zip)
[MobileAdvert bug.c3p.zip](https://github.com/Scirra/Construct-bugs/files/10945365/MobileAdvert.bug.c3p.zip)

## Steps to reproduce

1. click the load button to load the rewarded video
2. click the show button to show the video and let it finish
3. click the load button to load the rewarded video again
4. click the show button to show the video and see the error in the text log

if you click again the load button you'll be able to show the video.

## Observed result

In the preview, the application cancel the consecutive video

## Expected result
running the apk in a mobile device you'll be able to click the load and show button alternatively with no issue as expected.

## More details

**Affected browsers/platforms:** 
The problem appears on both Chrome and Firefox browsers.

**First affected release:** 
I have tested the problem with release R333
## System details

<details><summary>View details</summary>

PASTE HERE

</details>

**Comment by DiegoScirra** [comment]
When trying adverts in preview the test requests will alternate between one that works and one that doesn't so you are able to add events to handle the case in which an ad could not be obtained.

It looks like this behaviour is not documented anywhere, so I will add it to the manual.

**Comment by rraffo** [comment]
Good to know...
is this behavior defined by the plug-in or by google itself?
Is there a way to override this behavior to have all good videos?
Thanks

**Comment by DiegoScirra** [comment]
When a game runs in a web context then C3 is using Google's Ad placement API.

https://developers.google.com/ad-placement/docs/html5-game-structure

The behaviour is part of their library and can't be disabled as far as I know. In preview, C3 is just making mock ad requests and the feature exists to test the real possibility of a real request for an ad just failing. That can happen for a variety of reasons all of which are out of our control. In the case of failure the application shouldn't crash and instead handle the case transparently.

I imagine they added that because otherwise it is easy to forget about it and just implement your ads with out any kind of fail safe!

They are also giving you a hint on how they want you to use ads in your games. Basically they should be implemented in such a way that it doesn't matter if they are shown or not because you just don't know if the ad is going to be served or not.

**Comment by rraffo** [comment]
How can I handle the error of an unavailable video?
In the example, I load the video and I don't get any error when the load request is completed. I get a confirmation that the video is loaded and ready to be played. This is confusing me because I would expect that the error of an unavailable video would happen when the load request is performed and not when I show the video.
But it seems you are saying that the load of the video will occur in any case. Do I'll discover if it is available only when I try to show it?
thanks

**Comment by rraffo** [comment]
After showing the first load video the subsequent load of the rewarded video doesn't generate an error but actually triggers the "video is ready". 
Do I need instead to try to show the video to discover if the video exists or not?

**Comment by DiegoScirra** [comment]
To detect if there are any problems showing a rewarded ad you have to use the **"Failed To Load"** and **"Cancelled"** trigger events. There is a very subtle bug to watch out for though.

In mobile exports the **"Failed to Load"** trigger is associated with the **"Create"** actions and the **"Cancelled"** trigger is associated with the **"Show"** actions. So far so good.

In web exports things seem to be the same, but they are not. The library that Google provides has no notion of creating ads and showing them. You can only show them, that's it. We wanted the API of our plugin to be the same for both mobile and web, so some actions don't do anything on web exports and just complete successfully. That is why you are seeing the **"Create"** actions complete successfully in preview. The idea behind this decision is that in theory no changes need to be made to events no matter what is your export target.

There is a bug in web exports though, which causes the **"Cancelled"** trigger to be the only one you need to handle for both **"Create"** and **"Show"** actions. This will be fixed, but the fix will only become available when the next beta cycle starts, to avoid introducing potential regressions in the upcoming stable release.

In short, in web exports you don't know if an ad will be available until you try to show it. That is why the integration with your game should be done in such a way that progression is not tied to whether an ad was viewed or not, because you simply don't know if the ad will be available to begin with,

**Comment by DiegoScirra** [comment]
It's ok, I have known about this for a while (1+ year) but haven't had time to fix it. The thing to remember with ads is that what preview shows you is as if the game was deployed to a website. If you are targeting mobile you can use it to roughly test how your ad logic is going to work, but if you really want to see what it's going to be like, you need to try it out in an emulator with Android Studio / XCode, or better yet, a real device.

**Comment by Alaadel** [comment]
> To detect if there are any problems showing a rewarded ad you have to use the **"Failed To Load"** and **"Cancelled"** trigger events. There is a very subtle bug to watch out for though.
> 
> In mobile exports the **"Failed to Load"** trigger is associated with the **"Create"** actions and the **"Cancelled"** trigger is associated with the **"Show"** actions. So far so good.
> 
> In web exports things seem to be the same, but they are not. The library that Google provides has no notion of creating ads and showing them. You can only show them, that's it. We wanted the API of our plugin to be the same for both mobile and web, so some actions don't do anything on web exports and just complete successfully. That is why you are seeing the **"Create"** actions complete successfully in preview. The idea behind this decision is that in theory no changes need to be made to events no matter what is your export target.
> 
> There is a bug in web exports though, which causes the **"Cancelled"** trigger to be the only one you need to handle for both **"Create"** and **"Show"** actions. This will be fixed, but the fix will only become available when the next beta cycle starts, to avoid introducing potential regressions in the upcoming stable release.
> 
> Just make sure to use both triggers and have your application respond accordingly. You can use the **PlatformInfo** plugin to check what is the target platform and respond accordingly.
> 
> **NOTE:** You still need to use the **"Create"** actions in web exports, even if they don't do anything. Again, this has to do with making the events necessary for both mobile and web the same.
> 
> In short, in web exports you don't know if an ad will be available until you try to show it. That is why the integration with your game should be done in such a way that progression is not tied to whether an ad was viewed or not, because you simply don't know if the ad will be available to begin with,

Is this behavior/workaround documented please?
I have had similar issues in the past but couldn't pinpoint the cause or the solution.

**Comment by DiegoScirra** [comment]
Web exports only triggering the "Cancelled" event if there is a problem showing an ad is not documented because it is bug.

Adverts in preview alternating between valid and invalid requests is documented.

**Comment by DiegoScirra** [comment]
Unfortunately I haven't been able to properly try out the changes to fix this because my Admob/Adsense account was closed by Google. Now waiting to see if I can get it re-activated.

**Comment by rraffo** [comment]
Diego, thanks to keep this on your list.

**Comment by DiegoScirra** [comment]
The next beta will fix the issue of web exports always triggering the "Cancelled" event when using the different "Create" actions with the "show" property set to **true**
